test/e2e/tests/tcp-route.go

```diff
@@ -0,0 +1,245 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+//go:build e2e
+// +build e2e
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/gateway-api/conformance/utils/config"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, TCPRouteTest)
+}
+
+// GatewayRef is a tiny type for specifying an TCP Route ParentRef without
+// relying on a specific api version.
+type GatewayRef struct {
+	types.NamespacedName
+	listenerNames []*gatewayv1.SectionName
+}
+
+// NewGatewayRef creates a GatewayRef resource.  ListenerNames are optional.
+func NewGatewayRef(nn types.NamespacedName, listenerNames ...string) GatewayRef {
+	var listeners []*gatewayv1.SectionName
+
+	if len(listenerNames) == 0 {
+		listenerNames = append(listenerNames, "")
+	}
+
+	for _, listener := range listenerNames {
+		sectionName := gatewayv1.SectionName(listener)
+		listeners = append(listeners, &sectionName)
+	}
+	return GatewayRef{
+		NamespacedName: nn,
+		listenerNames:  listeners,
+	}
+}
+
+var TCPRouteTest = suite.ConformanceTest{
+	ShortName:   "TCPRoute",
+	Description: "Testing TCP Route",
+	Manifests:   []string{"testdata/tcp-route.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		t.Run("tcp-route-1", func(t *testing.T) {
+			ns := "gateway-conformance-infra"
+			routeNN := types.NamespacedName{Name: "tcp-app-1", Namespace: ns}
+			gwNN := types.NamespacedName{Name: "my-tcp-gateway", Namespace: ns}
+			gwAddr := GatewayAndTCPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, NewGatewayRef(gwNN), routeNN)
+			OkResp := http.ExpectedResponse{
+				Request: http.Request{
+					Path: "/",
+				},
+				Response: http.Response{
+					StatusCode: 200,
+				},
+				Namespace: ns,
+			}
+
+			// Send a request to an valid path and expect a successful response
+			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, OkResp)
```

Comment:
how is the destination port (8090, 8080) being computed ?

Reply:
As I understand following the [docs](https://gateway-api.sigs.k8s.io/guides/tcp/).  We have one Gateway resource(and two TCP listerns on port 8080 and 8090) and two TCPRoute resources. When e2e test is run gateways are created and requests are passed to the added services port for TCPRoute.  (Please correct me if I'm missing something here.)

Reply:
While checking for `WaitForGatewayAddress` function found that it only get's the [port of first listeners](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/kubernetes/helpers.go#L411) . In our case we have two listeners sets on (8080, 8090) , but as It is just getting the first port of the listener i.e `8080` . After running the e2e test I found that it is making request on same port for both listeners

```
tcp-route.go:83: Making GET request to http://172.18.255.205:8080/
tcp-route.go:102: Making GET request to http://172.18.255.205:8080/
```

Reply:
can you ensure traffic is sent to different ports

Reply:
Done, Now traffic is send to specific ports
```
tcp-route.go:78: Making GET request to http://172.18.255.205:8090/
tcp-route.go:60: Making GET request to http://172.18.255.205:8080/
```